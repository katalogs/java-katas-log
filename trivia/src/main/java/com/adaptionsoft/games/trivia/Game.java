package com.adaptionsoft.games.trivia;

import java.util.*;

import static com.adaptionsoft.games.trivia.Category.*;

public class Game {
    private final List<String> players = new ArrayList<>();
    private final int[] places = new int[6];
    private final boolean[] inPenaltyBox  = new boolean[6];
	private final int[] purses  = new int[6];
	private final DecksHolder decksHolder = new DecksHolder();
    private int currentPlayer = 0;
    private boolean isGettingOutOfPenaltyBox;
	private Board board = new Board();
    
    public Game(){


	}


	public boolean addPlayer(String playerName) {
	    players.add(playerName);
		final int playerNumber = players.size();
	    places[playerNumber] = 0;
	    purses[playerNumber] = 0;
	    inPenaltyBox[playerNumber] = false;
	    
	    System.out.println(playerName + " was added");
	    System.out.println("They are player number " + playerNumber);
		return true;
	}

	public void roll(int roll) {
		System.out.println(players.get(currentPlayer) + " is the current player");
		System.out.println("They have rolled a " + roll);
		
		if (isCurrentPlayerInPenaltyBox()) {
			if (isOdd(roll)) {
				isGettingOutOfPenaltyBox = true;
				System.out.println(players.get(currentPlayer) + " is getting out of the penalty box");
				moveCurrentPlayerAndPrintCategory(roll);
				askQuestion();
			} else {
				System.out.println(players.get(currentPlayer) + " is not getting out of the penalty box");
				isGettingOutOfPenaltyBox = false;
				}
		} else {
			moveCurrentPlayerAndPrintCategory(roll);
			askQuestion();
		}

	}

	private boolean isOdd(int roll) {
		return roll % 2 != 0;
	}

	private void moveCurrentPlayerAndPrintCategory(int roll) {
		places[currentPlayer] = (places[currentPlayer] + roll) % 12;
		System.out.println(players.get(currentPlayer)
				+ "'s new location is "
				+ places[currentPlayer]);
		System.out.println("The category is " + currentCategory());
	}

	private boolean isCurrentPlayerInPenaltyBox() {
		return inPenaltyBox[currentPlayer];
	}

	private void askQuestion() {
		System.out.println(decksHolder.drawQuestionByCategory(currentCategory()));
	}
	
	
	private Category currentCategory() {
		switch (places[currentPlayer]) {
			case 0:
			case 4:
			case 8:
				return Pop;
			case 1:
			case 5:
			case 9:
				return Science;
			case 2:
			case 6:
			case 10:
				return Sports;
			default:
				return Rock;
		}

	}

	public boolean wasCorrectlyAnswered() {
		if (isCurrentPlayerInPenaltyBox() && !isGettingOutOfPenaltyBox){
			goToNextPlayer();
			return true;
		}

		if (isGettingOutOfPenaltyBox) {
			System.out.println("Answer was correct!!!!");
		} else {
			System.out.println("Answer was corrent!!!!");
		}
		return rewardAndGoToNextPlayer();
	}

	private boolean rewardAndGoToNextPlayer() {
		rewardCurrentPlayer();
		boolean doesGameContinue = !didPlayerWin();
		goToNextPlayer();

		return doesGameContinue;
	}

	private void rewardCurrentPlayer() {
		purses[currentPlayer]++;
		System.out.println(players.get(currentPlayer)
				+ " now has "
				+ purses[currentPlayer]
				+ " Gold Coins.");
	}

	private void goToNextPlayer() {
		currentPlayer++;
		if (currentPlayer == players.size()) currentPlayer = 0;
	}

	public boolean wrongAnswer(){
		System.out.println("Question was incorrectly answered");
		sendCurrentPlayerToPenaltyBox();
		goToNextPlayer();
		return true;
	}

	private void sendCurrentPlayerToPenaltyBox() {
		System.out.println(players.get(currentPlayer)+ " was sent to the penalty box");
		inPenaltyBox[currentPlayer] = true;
	}


	private boolean didPlayerWin() {
		return purses[currentPlayer] == 6;
	}
}
