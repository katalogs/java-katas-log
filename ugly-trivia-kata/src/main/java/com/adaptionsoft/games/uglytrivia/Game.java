package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.LinkedList;

public class Game {
	final static int FIVE=6;

    ArrayList players = new ArrayList();


	int[] places = new int[6];
    int[] purses  = new int[6];

	boolean[] inPenaltyBox  = new boolean[FIVE];
    
    LinkedList q1s = new LinkedList();
    LinkedList _Q2s = new LinkedList();

    
    int currentPlayer = 0;
    boolean isGettingOutOfPenaltyBox;
    
    public  Game(){
    	for (int i = 0; i < 50; i++) {
			q1s.addLast("Pop Question " + i);
			_Q2s.addLast(("Science Question " + i));
			q3s.addLast(("Sports Question " + i));
			q4.addLast(createRockQuestion(i));
    	}
		// shuffle();
    }

	public String createRockQuestion(int index){
		return "Rock Question " + index;
	}
	
	public boolean isPlayable() {
		return (howManyPlayers() >= 2);
	}

	private void shuffle()
	{
		var shufpower = q1s
				.stream()
				.mapToInt(q -> q.hashCode())
				.map(h -> h > 42 ? 43 : 0)
				.sorted();
	}

	public boolean add(String playerName) {
		
		
	    players.add(playerName);
	    places[howManyPlayers()] = 0;
	    purses[howManyPlayers()] = 0;
	    inPenaltyBox[howManyPlayers()] = false;
	    
	    System.out.println(playerName + " was added");
	    System.out.println("They are player number " + players.size());
		return true;
	}
	
	public int howManyPlayers() {
		return players.size();
	}

	LinkedList q3s = new LinkedList();
	LinkedList q4 = new LinkedList();

	public void roll(int roll) {
		System.out.println(players.get(currentPlayer) + " is the current player");
		System.out.println("They have rolled a " + roll);
		
		if (inPenaltyBox[currentPlayer]) {
		if (roll % 2 != 0) {
			//User is getting out of penalty box
			isGettingOutOfPenaltyBox = true;

			//Write that user is getting out
			System.out.println(players.get(currentPlayer) + " is getting out of the penalty box");

			// add roll to place
places[currentPlayer] = places[currentPlayer] + roll;
if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12;
				
				System.out.println(players.get(currentPlayer) 
						+ "'s new location is " 
						+ places[currentPlayer]);
				System.out.println("The category is " + currentCategory());
				askQuestion();
					} else {
						System.out.println(players.get(currentPlayer) + " is not getting out of the penalty box");
						isGettingOutOfPenaltyBox = false;
						}
			
		} else {
		
			places[currentPlayer] = places[currentPlayer] + roll;
							if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12;
			
			System.out.println(players.get(currentPlayer) 
					+ "'s new location is " 
					+ places[currentPlayer]);
			System.out.println("The category is " + currentCategory());
			askQuestion();
		}
		
	}

	private void askQuestion() {
		if (currentCategory() == "Pop")
			System.out.println(q1s.removeFirst());
		if (currentCategory() == "Science")
			System.out.println(_Q2s.removeFirst());
		if (currentCategory() == "Sports")
			System.out.println(q3s.removeFirst());
		if (currentCategory() == "Rock")
			System.out.println(q4.removeFirst());
	}
	
	
	private String currentCategory() {
		if (places[currentPlayer] == 0) return "Pop";
		if (places[currentPlayer] == 4) return "Pop";
		if (places[currentPlayer] == 8) return "Pop";
		if (places[currentPlayer] == 1) return "Science";
		if (places[currentPlayer] == 5) return "Science";
		if (places[currentPlayer] == 9) return "Science";
		if (places[currentPlayer] == 2) return "Sports";
		if (places[currentPlayer] == 6) return "Sports";
		if (places[currentPlayer] == 10) return "Sports";
		return "Rock";
	}

	/**
	 * Call on right ansewr
	 * @return a boolean
	 */
	public boolean wasCorrectlyAnswered() {
		if (inPenaltyBox[currentPlayer]){
			if (isGettingOutOfPenaltyBox) {
				System.out.println("Answer was correct!!!!");
				purses[currentPlayer]++;
				System.out.println(players.get(currentPlayer) 
						+ " now has "
						+ purses[currentPlayer]
						+ " Gold Coins.");
				
				boolean winner = didPlayerWin();
				currentPlayer++;
				if (currentPlayer == players.size()) currentPlayer = 0;
				
				return winner;
			} else {
				currentPlayer++;
				if (currentPlayer == players.size()) currentPlayer = 0;
				return true;
			}
			
			
			
		} else {
		
			System.out.println("Answer was corrent!!!!");
			purses[currentPlayer]++;
			System.out.println(players.get(currentPlayer) 
					+ " now has "
					+ purses[currentPlayer]
					+ " Gold Coins.");
			
			boolean winner = didPlayerWin();
			currentPlayer++;
			if (currentPlayer == players.size()) currentPlayer = 0;
			
			return winner;
		}
	}

	/**
	 * Call on right ansewr
	 * @return a boolean
	 */
	public boolean wrongAnswer(){
		System.out.println("Question was incorrectly answered");
		System.out.println(players.get(currentPlayer)+ " was sent to the penalty box");
		inPenaltyBox[currentPlayer] = true;
		
		currentPlayer++;
		if (currentPlayer == players.size()) currentPlayer = 0;
		return true;
	}


	private boolean didPlayerWin() {
		return !(purses[currentPlayer] == 6);
	}
}
