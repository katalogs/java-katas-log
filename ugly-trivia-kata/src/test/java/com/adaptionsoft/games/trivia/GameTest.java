package com.adaptionsoft.games.trivia;


import com.adaptionsoft.games.uglytrivia.Game;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.*;

import static org.approvaltests.Approvals.verify;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GameTest {
    private ByteArrayOutputStream out;

    @BeforeEach
    public void setUpOutput() {
        out = new ByteArrayOutputStream();
        System.setOut(new PrintStream(out));
    }

    @Test
    void test1() {
        var game = new Game();
        game.add("Bart");
        game.roll(12);
        game.wrongAnswer();
        game.roll(2);
        game.roll(13);
        game.wasCorrectlyAnswered();
        game.roll(13);

        verify(out.toString());
    }

    @Test
    void test2() {
        var game = new Game();
        game.add("Bart");
        game.add("Lisa");
        game.roll(1);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();

        verify(out.toString());
    }

    @Test
    void test3() {
        var game = new Game();
        game.add("Bart");
        game.add("Lisa");
        game.roll(1);
        game.wrongAnswer();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();
        game.roll(2);
        game.wasCorrectlyAnswered();

        verify(out.toString());
    }
}
