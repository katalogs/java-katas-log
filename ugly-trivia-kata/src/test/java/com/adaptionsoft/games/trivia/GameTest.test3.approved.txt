Bart was added
They are player number 1
Lisa was added
They are player number 2
Bart is the current player
They have rolled a 1
Bart's new location is 1
The category is Science
Science Question 0
Question was incorrectly answered
Bart was sent to the penalty box
Lisa is the current player
They have rolled a 2
Lisa's new location is 2
The category is Sports
Sports Question 0
Answer was corrent!!!!
Lisa now has 1 Gold Coins.
Bart is the current player
They have rolled a 2
Bart is not getting out of the penalty box
Lisa is the current player
They have rolled a 2
Lisa's new location is 4
The category is Pop
Pop Question 0
Answer was corrent!!!!
Lisa now has 2 Gold Coins.
Bart is the current player
They have rolled a 2
Bart is not getting out of the penalty box
Lisa is the current player
They have rolled a 2
Lisa's new location is 6
The category is Sports
Sports Question 1
Answer was corrent!!!!
Lisa now has 3 Gold Coins.
Bart is the current player
They have rolled a 2
Bart is not getting out of the penalty box
Lisa is the current player
They have rolled a 2
Lisa's new location is 8
The category is Pop
Pop Question 1
Answer was corrent!!!!
Lisa now has 4 Gold Coins.
Bart is the current player
They have rolled a 2
Bart is not getting out of the penalty box
Lisa is the current player
They have rolled a 2
Lisa's new location is 10
The category is Sports
Sports Question 2
Answer was corrent!!!!
Lisa now has 5 Gold Coins.
